scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java

```diff
@@ -0,0 +1,247 @@
+  private Duration lastSimResultsEnd = Duration.ZERO;
+
+  //map from activity start time to serialized activity and its name
+  private List<SimulatedActivity> activitiesInserted = new ArrayList<>();
+
+  record SimulatedActivity(Duration start, SerializedActivity activity, String name) implements Comparable<SimulatedActivity> {
+    @Override
+    public int compareTo(@NotNull final SimulatedActivity o) {
+      return start.compareTo(o.start);
+    }
+  }
+
+  public IncrementalSimulationDriver(MissionModel<?> missionModel){
+    this.missionModel = missionModel;
+    initSimulation();
+  }
+
+  private void initSimulation(){
+    actToId = new HashMap<>();
+    lastSimResults = null;
+    this.engine = new SimulationEngine();
+    activitiesInserted.clear();
+
+    /* The top-level simulation timeline. */
+    this.timeline = new TemporalEventSource();
+    areLastSimResultsDirty = false;
+    this.cells = new LiveCells(timeline, missionModel.getInitialCells());
+    /* The current real time. */
+    curTime = Duration.ZERO;
+
+    // Begin tracking all resources.
+    for (final var entry : missionModel.getResources().entrySet()) {
+      final var name = entry.getKey();
+      final var resource = entry.getValue();
+      engine.trackResource(name, resource, curTime);
+    }
+
+    // Start daemon task(s) immediately, before anything else happens.
+    {
+      final var daemon = engine.initiateTaskFromSource(missionModel::getDaemon);
+      final var commit = engine.performJobs(Set.of(SimulationEngine.JobId.forTask(daemon)),
+                                            cells, curTime, Duration.MAX_VALUE, missionModel);
+      timeline.add(commit);
+    }
+  }
+
+  public void simulateActivity(SerializedActivity activity, Duration startTime, String nameAct){
+    var testAct = new SimulatedActivity(startTime, activity, nameAct);
+    if(startTime.noLongerThan(curTime)){
+      var toBeInserted = new ArrayList<>(activitiesInserted);
+      toBeInserted.add(testAct);
+      initSimulation();
+      simulateManyActs(toBeInserted);
+    } else {
+      simulateAct(testAct);
+    }
+  }
+
+
+  /**
+   * Get the simulation results from the Duration.ZERO to the current simulation time point
+   * @return the simulation results
+   */
+  public SimulationResults getSimulationResults(){
+    return getSimulationResultsUntil(curTime);
+  }
+
+  /**
+   * Get the simulation results from the Duration.ZERO to a specified end time point.
+   * The provided simulation results might cover more than the required time period.
+   * @return the simulation results
+   */  public SimulationResults getSimulationResultsUntil(Duration endTime){
+    //if previous results cover a bigger period, we return do not regenerate
+    if(areLastSimResultsDirty || lastSimResults == null || endTime.longerThan(lastSimResultsEnd)) {
+      lastSimResults = engine.computeResults(
+          engine,
+          Instant.now(),
+          endTime,
+          new HashMap<>(),
```

Comment:
I'm a little concerned about this empty `HashMap`. This map is used by `computeResults` to replace engine-internal task IDs with the activity IDs sourced from the original plan. Normally we pass `controlTask.extractTaskToPlannedDirective()` here, but this class doesn't hold onto the `controlTask`s it constructs.

Reply:
I have replaced this empty map by the `taskToDirective` map now locally kept by the `IncrementalSimulationDriver`.